### PR TITLE
DQt2 logging dockable window

### PR DESCRIPTION
**Pull request by rukai**
Adds a log in the form of a docked window.
Colors were pulled from my terminal, let me know of better color schemes.
I cleaned up the LogManager code a bit, to make it more obvious what its doing and moved code from the GUI into LogManager.
I used a check box to toggle the log window on and off, alternatively I could have a standard menu option and let the user close it via the x button.
Log configuration is coming later.

screenshot

[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3416)

### CODE REVIEW

File: Source/Core/DolphinQt2/Config/LogDialog.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#include <QObject>
+
+#include "Common/IniFile.h"
+#include "Common/Logging/LogManager.h"
+#include "DolphinQt2/LogViewer.h"
+
+class LogDialog final : public QDialog
+{
+	Q_OBJECT
+	public:
```
Comment (spxtr) [inline Source/Core/DolphinQt2/Config/LogDialog.h]: no indent

File: Source/Core/DolphinQt2/LogViewer.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+	QTimer *timer = new QTimer(this);
+	connect(timer, SIGNAL(timeout()), this, SLOT(displayLog()));
+	timer->start(200);
+}
+
+LogViewer::~LogViewer()
+{
+	LogManager* m_LogManager = LogManager::GetInstance();
+	if(m_LogManager)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.cpp]: space after if

File: Source/Core/DolphinQt2/LogViewer.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		m_LogManager->CloseWindow();
+}
+
+void LogViewer::Log(LogTypes::LOG_LEVELS level, const char* msg)
+{
+	std::lock_guard<std::mutex> lk(m_LogEntry);
+	QString qmsg = QString::fromUtf8(msg).trimmed();
+	msgQueue.push(std::make_pair(level, qmsg));
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.cpp]:
```cpp
msgQueue.emplace(level, qmsg)
```

File: Source/Core/DolphinQt2/LogViewer.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+void LogViewer::displayLog()
+{
+	std::lock_guard<std::mutex> lk(m_LogEntry);
+	bool scroll = false;
+	// This function runs on the main gui thread, and needs to finish in a finite time otherwise
+	// the GUI will lock up, which could be an issue if new messages are flooding in faster than
+	// this function can render them to the screen.
+	// So we limit the number of times run via this for loop.
+	for(int i = 0; i < 100; i++)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.cpp]: space after for

File: Source/Core/DolphinQt2/LogViewer.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	{
+		if(msgQueue.empty())
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.cpp]: space after if

File: Source/Core/DolphinQt2/LogViewer.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+			setTextColor(QColor(70, 70, 70, 70)); //grey
+			break;
+		default:
+			setTextColor(QColor(221, 221, 221)); //white
+			break;
+		}
+		insertPlainText(qmsg.right(qmsg.length()-9));
+		scroll = true;
+	}
+	if(scroll)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.cpp]: space after if

File: Source/Core/DolphinQt2/LogViewer.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	Q_OBJECT
+public:
+	explicit LogViewer(QWidget* parent);
+	~LogViewer();
+	void Log(LogTypes::LOG_LEVELS, const char* msg) override;
+private slots:
+	void displayLog();
+private:
+	std::mutex m_LogEntry;
+	std::queue <std::pair<LogTypes::LOG_LEVELS, QString>> msgQueue;
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.h]: No space after queue

File: Source/Core/DolphinQt2/LogViewer.h [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <mutex>
+#include <QObject>
+#include <QString>
+#include <QTextEdit>
+#include <queue>
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.h]: standard library headers should be included first.

File: Source/Core/DolphinQt2/LogViewer.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+// Refer to the license.txt file included.
+
+#include <mutex>
+#include <QScrollBar>
+#include <QString>
+#include <QTextEdit>
+#include <QTimer>
+#include <queue>
+#include <QWidget>
+#include <utility>
```
Comment (lioncash) [inline Source/Core/DolphinQt2/LogViewer.cpp]: Standard library headers should be included first.

File: Source/Core/DolphinQt2/LogViewer.h (hunk 8 above)
Comment (rukai) [inline Source/Core/DolphinQt2/LogViewer.h]: That goes against the linter. Is this what we want? Should I change the linter?

File: Source/Core/Common/Logging/LogManager.cpp [hunk 10]
```diff
@@ -85,22 +85,69 @@ LogManager::LogManager()
...
+	options->Get("Verbosity", &verbosity, 0);
+
+	// Ensure the verbosity level is valid
+	if (verbosity < 1)
+		verbosity = 1;
+	if (verbosity > MAX_LOGLEVEL)
+		verbosity = MAX_LOGLEVEL;
+
+	// Get the logger output settings from the config ini file.
+	bool m_writeFile, m_writeConsole;
```
Comment (JosJuice) [inline Source/Core/Common/Logging/LogManager.cpp]: These variables are local, so they shouldn't use the `m_` prefix.

Comment (JosJuice) [inline Source/Core/Common/Logging/LogManager.cpp]: Also, I don't think DolphinWX's `m_writeFile` and `m_writeConsole` are used anymore, so you can remove the declarations of those.

**Comment by BhaaLseN** [comment]
Not really happy with having `LogManager::XxxWindow` methods that are only there for callers with windows/tabs/popups/dialogs/whatever. IMO that belongs in some other place, like a UI Common class or whatever.
Can't think of any other ways that makes it generic enough so it can stay there, so you might aswell ignore my comment if other devs disagree or think its fine to keep it like that.

**Comment by spxtr** [comment]
For the future, I recommend making changes outside of DQt2 as a separate PR, or at least as a separate commit.

**Comment by RisingFog** [comment]
Any update for this PR? It'll need to be rebased, regardless.

**Comment by leoetlino** [comment]
Closing (superseded by #5730)

**Pull request closed without merging**